core/src/apps/cardano/address.py

```diff
@@ -68,22 +81,66 @@ def _has_shelley_address_header(address: bytes) -> bool:
     )
 
 
+def _validate_output_shelley_address(
+    address_str: str, address_bytes: bytes, network_id: int
+) -> None:
+    address_type = _get_address_type(address_bytes)
+    if address_type == CardanoAddressType.REWARD:
+        raise INVALID_ADDRESS
+
+    _validate_output_address_bech32_hrp(address_str, address_type, network_id)
+    _validate_address_network_id(address_bytes, network_id)
+
+
+def _validate_output_address_bech32_hrp(
+    address_str: str, address_type: int, network_id: int
+) -> None:
+    valid_hrp = _get_bech32_hrp_for_address(address_type, network_id)
+    bech32_hrp = bech32.get_hrp(address_str)
+
+    if valid_hrp != bech32_hrp:
+        raise INVALID_ADDRESS
+
+
+def _get_bech32_hrp_for_address(address_type: int, network_id: int) -> str:
+    if address_type == CardanoAddressType.BYRON:
+        raise ValueError("Byron address uses base58 encoding")
+
+    if address_type == CardanoAddressType.REWARD:
+        if network_id == network_ids.MAINNET:
```

Comment:
as we spoke offline on Friday, is there a single "mainnet" network id to compare against? Isn't there only a single Testnet id (0x00)? Shoultn't in that case be there the reverse condition, i.e. `network_id != network_ids.TESTNET` ?

Reply:
also, are we definitely sure about current mainnet network id being 1? If not, I would rather refrain from asserting in `_validate_network_info` that network id is 1 if and only if protocol magic is the mainnet one and rather just assert that network id mustn't be 0 if protocol magic is mainnet - this would work even if IOHK chose a different network id last minute or if new mainnets are launched later on

Reply:
I've extracted the check into a method in `network_ids.py` and it now checks that the network id is not a testnet one as you suggested. Thanks!